src/Discord.Net/State/IStateProvider.cs

```diff
@@ -0,0 +1,36 @@
+// todo: impl, docs
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Discord
+{
+    public interface IStateProvider
+    {
+        // we'll need to extend this as the client is implemented
+
+        ValueTask<IUser> GetUserAsync(ulong id, StateBehavior stateBehavior, RequestOptions? options);
+        // review: IReadOnlyCollection is proper collection type here?
+        ValueTask<IReadOnlyCollection<IUser>> GetUsersAsync(StateBehavior stateBehavior, RequestOptions? options);
```

Comment:
Yes. As long as the immutable collections aren't used (they're very slow and memory heavy).

Reply:
Not sure I agree with that assessment of immutables, in general, but I definitely agree that `IReadOnlyCollection` is the best bet here. Immutables would be appropriate if you KNOW your consumer is going to be doing a lot of collection manipulation, and also wants immutability.

The other option would be `IReadOnlyList` but do we really expect consumers to want indexing by position?